compiler-core/src/type_/tests/expression.rs

```diff
@@ -0,0 +1,164 @@
+            ("erlang_only_1", Implementations::erlang()),
+            ("erlang_only_2", Implementations::erlang()),
+        ],
+    );
+}
+
+#[test]
+pub fn externals_only_function() {
+    assert_targets!(
+        r#"
+@external(erlang, "foo", "bar")
+@external(javascript, "foo", "bar")
+pub fn all_externals_1() -> Int
+
+pub fn all_externals_2() { all_externals_1() * 2 }
+"#,
+        vec![
+            ("all_externals_1", Implementations::all_externals()),
+            ("all_externals_2", Implementations::all_externals())
+        ],
+    );
+}
+
+#[test]
+pub fn externals_with_pure_gleam_body() {
+    assert_targets!(
+        r#"
+@external(javascript, "foo", "bar")
+pub fn javascript_external_and_pure_body() -> Int { 1 + 1 }
+
+@external(erlang, "foo", "bar")
+pub fn erlang_external_and_pure_body() -> Int { 1 + 1 }
+
+pub fn pure_gleam() {
+  javascript_external_and_pure_body() + erlang_external_and_pure_body()
+}
+"#,
+        vec![
+            (
+                "erlang_external_and_pure_body",
+                Implementations {
+                    gleam: true,
+                    erlang: true,
+                    javascript: false
+                }
+            ),
+            (
+                "javascript_external_and_pure_body",
+                Implementations {
+                    gleam: true,
+                    erlang: false,
+                    javascript: true
+                }
+            ),
+            (
+                "pure_gleam",
+                Implementations {
+                    gleam: true,
+                    erlang: true,
+                    javascript: true
+                }
+            )
+        ],
+    );
+}
+
+#[test]
+pub fn erlang_external_with_javascript_body() {
+    assert_targets!(
+        r#"
+@external(javascript, "foo", "bar")
+pub fn javascript_only() -> Int
+
+@external(erlang, "foo", "bar")
+pub fn erlang_external_and_javascript_body() -> Int { javascript_only() }
+
+pub fn all_externals() -> Int { erlang_external_and_javascript_body() }
+"#,
+        vec![
+            ("all_externals", Implementations::all_externals()),
+            (
+                "erlang_external_and_javascript_body",
+                Implementations::all_externals()
+            ),
+            ("javascript_only", Implementations::javascript()),
+        ],
+    );
+}
+
+#[test]
+pub fn javascript_external_with_erlang_body() {
+    assert_targets!(
+        r#"
+@external(erlang, "foo", "bar")
+pub fn erlang_only() -> Int
+
+@external(javascript, "foo", "bar")
+pub fn javascript_external_and_erlang_body() -> Int { erlang_only() }
+
+pub fn all_externals() -> Int { javascript_external_and_erlang_body() }
+"#,
+        vec![
+            ("all_externals", Implementations::all_externals()),
+            ("erlang_only", Implementations::erlang()),
+            (
+                "javascript_external_and_erlang_body",
+                Implementations::all_externals()
+            ),
+        ],
+    );
+}
```

Comment:
Can we have a test for one that ends up with all false?

Reply:
Depending on how what we un up agreeing [here](https://github.com/gleam-lang/gleam/pull/2541#discussion_r1457677570) it might or might not be possible to have Implementations end up with all fields set to false because the compiler would emit an error earlier.
- If we go along with the current implementation than it's not possible because the only way to get it would be by having a function with no body and no annotations (and that would result in a compilation error)
- Otherwise it might be possible to have that

Reply:
It could be a test that results in an error, that's cool, we just need to make sure it covers all the paths.